File: backend/booking-service/src/main/java/io/internview/booking_service/web/dto/PageResponse.java
package io.internview.booking_service.web.dto;

import java.util.List;

import org.springframework.data.domain.Page;

import lombok.Builder;
import lombok.Value;

@Value
@Builder
public class PageResponse<T> {

	List<T> items;
	int page;
	int size;
	long totalElements;
	int totalPages;
	boolean hasNext;
	boolean hasPrevious;

	public static <T> PageResponse<T> from(Page<T> page) {
		return PageResponse.<T>builder()
			.items(page.getContent())
			.page(page.getNumber())
			.size(page.getSize())
			.totalElements(page.getTotalElements())
			.totalPages(page.getTotalPages())
			.hasNext(page.hasNext())
			.hasPrevious(page.hasPrevious())
			.build();
	}
}

File: mobile/internview/lib/core/models/page_response.dart
class PageResponse<T> {
  PageResponse({
    required this.items,
    required this.page,
    required this.size,
    required this.totalElements,
    required this.totalPages,
    required this.hasNext,
    required this.hasPrevious,
  });

  final List<T> items;
  final int page;
  final int size;
  final int totalElements;
  final int totalPages;
  final bool hasNext;
  final bool hasPrevious;

  factory PageResponse.fromJson(
    Map<String, dynamic> json,
    T Function(Object? e) itemFromJson,
  ) {
    final raw = json['items'];
    final list = raw is List ? raw : const [];
    return PageResponse(
      items: list.map(itemFromJson).toList(),
      page: (json['page'] as num?)?.toInt() ?? 0,
      size: (json['size'] as num?)?.toInt() ?? 0,
      totalElements: (json['totalElements'] as num?)?.toInt() ?? 0,
      totalPages: (json['totalPages'] as num?)?.toInt() ?? 0,
      hasNext: json['hasNext'] as bool? ?? false,
      hasPrevious: json['hasPrevious'] as bool? ?? false,
    );
  }
}
